### DISCUSSION TITLE
Why `Hash` instead of `Digest`?

### DISCUSSION THREAD

**Discussion by skaunov** [arkworks-rs/algebra#706] [General]
Why was it chosen for `Bls12` to implement [`Hash`](https://docs.rs/ark-ec/0.4.2/ark_ec/models/bls12/struct.Bls12.html#impl-Hash-for-Bls12%3CP%3E)? It seems to me that `Hash` doesn't suppose a cryptographic function and mostly used for `HashMap` and such things; while cryptohash are defined by something like `Digest` trait.

**Comment by burdges** [comment]
Bls12 is a ZST, so hashing it does nothing.  It likely supports `Hash` to help derive `Hash` elsewhere.

`Hasher` is the non-cryptographic analog of `Digest`.   Afaik `Digest` only handles bytes, so no analog of `Hash` exists for `Digest`

**Reply by skaunov** [reply]
Thank you! Sorry in advance if I ask anything stupid...

AFAIU its `Hash` is used so that user could hash something to curve for further pairing. Now I wonder if it's possible it was intended to be `HashToCurve` (maybe indirectly), but ended up this way... :thinking: Did you ever encounter `arkworks-rs` `Hash` in the wild? Now I want to see one! :eyeglasses: 
> Bls12 is a ZST, so hashing it does nothing. It likely supports Hash to help derive Hash elsewhere.

Isn't it on purpose? I mean representing anything as bytearray enables few usefull properties for cryptography, and isn't that restrictive. (Just not needed outside of crypto context, but I feel like `Bls12` very much *in*.)
> Hasher is the non-cryptographic analog of Digest. Afaik Digest only handles bytes, so no analog of Hash exists for Digest

**Reply by skaunov** [reply]
On the second look it really looks like to be able to `HashMap` points and elements if one would need this: <https://docs.rs/ark-ec/0.4.2/ark_ec/?search=Hasher>.

I would add that hashing to curve is currently so confusing I was quite sure it doesn't work yet until I found some code from a fellow who actually run it. I mean having it more clear would prevent this discussion. X) At least some note at `hashing` module...
